node/collation-generation/src/lib.rs

```diff
@@ -0,0 +1,222 @@
+			Ok(Signal(ActiveLeaves(ActiveLeavesUpdate { activated, .. }))) => {
+				// follow the procedure from the guide
+				if let Some(ref config) = self.config {
+					if let Err(err) = handle_new_activations(config, &activated, ctx).await {
+						log::warn!(target: "collation_generation", "failed to handle new activations: {:?}", err);
+						return true;
+					};
+				}
+				false
+			}
+			Ok(Signal(Conclude)) => true,
+			Ok(Communication {
+				msg: CollationGenerationMessage::Initialize(config),
+			}) => {
+				if self.config.is_some() {
+					log::warn!(target: "collation_generation", "double initialization");
+					true
+				} else {
+					self.config = Some(config);
+					false
+				}
+			}
+			Ok(Signal(BlockFinalized(_))) => false,
+			Err(err) => {
+				log::error!(target: "collation_generation", "error receiving message from subsystem context: {:?}", err);
+				true
+			}
+		}
+	}
+}
+
+impl<Context> Subsystem<Context> for CollationGenerationSubsystem
+where
+	Context: SubsystemContext<Message = CollationGenerationMessage>,
+{
+	fn start(self, ctx: Context) -> SpawnedSubsystem {
+		let subsystem = CollationGenerationSubsystem { config: None };
+
+		let future = Box::pin(subsystem.run(ctx));
+
+		SpawnedSubsystem {
+			name: "CollationGenerationSubsystem",
+			future,
+		}
+	}
+}
+
+#[derive(Debug, derive_more::From)]
+enum Error {
+	#[from]
+	Subsystem(SubsystemError),
+	#[from]
+	OneshotRecv(oneshot::Canceled),
+	#[from]
+	Runtime(RuntimeApiError),
+	#[from]
+	Util(util::Error),
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+async fn handle_new_activations<Context: SubsystemContext>(
+	config: &CollationGenerationConfig,
+	activated: &[Hash],
+	ctx: &mut Context,
+) -> Result<()> {
+	// follow the procedure from the guide:
+	// https://w3f.github.io/parachain-implementers-guide/node/collators/collation-generation.html
+
+	for relay_parent in activated.iter().copied() {
+		let global_validation_data = request_global_validation_data_ctx(relay_parent, ctx)
+			.await?
+			.await??;
+
+		let availability_cores = request_availability_cores_ctx(relay_parent, ctx)
+			.await?
+			.await??;
+
+		for core in availability_cores {
+			let (scheduled_core, assumption) = match core {
+				CoreState::Scheduled(scheduled_core) => {
+					(scheduled_core, OccupiedCoreAssumption::Free)
+				}
+				CoreState::Occupied(_occupied_core) => {
+					// TODO: https://github.com/paritytech/polkadot/issues/1573
+					continue
+				}
+				_ => continue,
+			};
+
+			let local_validation_data = match request_local_validation_data_ctx(
+				relay_parent,
+				scheduled_core.para_id,
+				assumption,
+				ctx,
+			)
+			.await?
+			.await??
+			{
+				Some(local_validation_data) => local_validation_data,
+				None => continue,
+			};
+
+			let validation_data_hash =
+				validation_data_hash(&global_validation_data, &local_validation_data);
+
+			let collation = (config.collator)(&global_validation_data, &local_validation_data).await;
```

Comment:
This and everything below it might be better if spawned as a sub-task. My rationale is that we can parallelize the loop over all the new activations and don't have to block the main collator task on collating, which might be the case if we receive two blocks on different relay-chain forks with a small time difference.

Another way to achieve a similar result would be for this function (`handle_new_activations`) to return a `FuturesUnordered` or a `JoinAll`. But that would still block the `handle_incoming` loop, which I think is best avoided.